crates/core/tedge_api/src/workflow/supervisor.rs

```diff
@@ -52,22 +66,159 @@ impl WorkflowSupervisor {
+            });
+        };
+        match GenericCommandState::from_command_message(message)? {
+            None => {
+                // The command has been cleared
+                self.commands.remove(&message.topic.name);
+                Ok(None)
+            }
+            Some(command_state) if command_state.status == "init" => {
+                // This is a new command request
+                self.commands.insert(command_state.clone())?;
+                Ok(Some(command_state))
+            }
+            Some(_) => {
+                // Ignore command updates published over MQTT
+                //
+                // TODO: There is exception here - not implemented yet:
+                //       when a step is delegated to an external process,
+                //       this process will notify the outcome of its action over MQTT,
+                //       and the agent will have then to react on this message.
+                Ok(None)
+            }
+        }
+    }
+
+    /// Return the action to be performed on a given command state
+    pub fn get_action(
+        &self,
+        command_state: &GenericCommandState,
+    ) -> Result<OperationAction, WorkflowExecutionError> {
+        let Some(operation_name) = command_state.operation() else {
+            return Err(WorkflowExecutionError::InvalidCmdTopic {
+                topic: command_state.topic.name.clone(),
+            });
+        };
+
         self.workflows
-            .get(operation)
-            .ok_or_else(|| WorkflowExecutionError::UnknownOperation {
-                operation: operation.into(),
+            .get(&operation_name.as_str().into())
+            .ok_or(WorkflowExecutionError::UnknownOperation {
+                operation: operation_name,
             })
-            .and_then(|workflow| OperationWorkflow::get_operation_current_action(workflow, status))
+            .and_then(|workflow| workflow.get_action(command_state))
+    }
+
+    /// Update the state of the command board on reception of new state for a command
+    ///
+    /// Return the next CommandRequest state if any is required.
+    pub fn apply_internal_update(
+        &mut self,
+        new_command_state: GenericCommandState,
+    ) -> Result<(), WorkflowExecutionError> {
+        self.commands.update(new_command_state)
+    }
+
+    /// Resume the given command when the agent is restarting after an interruption
+    pub fn resume_command(
+        &self,
+        _timestamp: &Timestamp,
+        command: &GenericCommandState,
+    ) -> Option<GenericCommandState> {
+        let Ok(action) = self.get_action(command) else {
+            return None;
+        };
+
+        match action {
+            OperationAction::AwaitingAgentRestart { on_success, .. } => {
+                Some(command.clone().update(on_success))
+            }
+
+            _ => {
+                // TODO: Use the timestamp to filter out action pending since too long
+                Some(command.clone())
```

Comment:
Is this safe? Shouldn't we be returning the `on_failed` target of these commands, as we can't expect to resume commands just like that from their current states, if they were interrupted in the middle of their execution last time, right? I mean, for it to work, we should be sure that the `script` actions provided by the users are also resumable even after a first incomplete attempt, right?

Reply:
After making that comment, I realised that it has been like this since we started supporting state transitions using MQTT retained messages, which would have behaved exactly the same way, when the same retained messages are re-delivered to the agent post-restart. I'm just not sure if our users would be aware of the impact: **"their scripts should essentially be resumable"**

Reply:
Very good point. I opt for this implementation precisely for the reason you give: "this  been like this since we started supporting state transitions using MQTT retained messages".

And yes the scripts should be resumable for that to work. We can add an option to "fail on retry".